pydantic_ai_slim/pydantic_ai/agent.py

```diff
@@ -1058,3 +1157,269 @@ def _prepare_result_schema(
+        ],
+    ) -> (
+        BaseNode[
+            _agent_graph.GraphAgentState,
+            _agent_graph.GraphAgentDeps[AgentDepsT, Any],
+            FinalResult[ResultDataT],
+        ]
+        | End[FinalResult[ResultDataT]]
+    ):
+        """Manually drive the agent run by passing in the node you want to run next.
+
+        This lets you inspect or mutate the node before continuing execution, or skip certain nodes
+        under dynamic conditions. The agent run should be stopped when you return an [`End`][pydantic_graph.nodes.End]
+        node.
+
+        Example:
+        ```python
+        from pydantic_ai import Agent
+        from pydantic_graph import End
+
+        agent = Agent('openai:gpt-4o')
+
+        async def main():
+            nodes = []
+            with agent.iter('What is the capital of France?') as agent_run:
+                # The first node can be retrieved via __anext__(), or you might already have it.
+                next_node = await agent_run.__anext__()
+                while not isinstance(next_node, End):
+                    next_node = await agent_run.next(next_node)
+                    nodes.append(next_node)
+                # Once `next_node` is an End, we've finished:
+                print(nodes)
+                '''
+                [
+                    HandleResponseNode(
+                        model_response=ModelResponse(
+                            parts=[TextPart(content='Paris', part_kind='text')],
+                            model_name='function:model_logic',
+                            timestamp=datetime.datetime(...),
+                            kind='response',
+                        )
+                    ),
+                    End(data=FinalResult(data='Paris', tool_name=None)),
+                ]
+                '''
+                print('Final result:', agent_run.final_result.data)
+                #> Final result: Paris
+        ```
+
+        Args:
+            node: The node to run next in the graph.
+
+        Returns:
+            The next node returned by the graph logic, or an [`End`][pydantic_graph.nodes.End] node if
+            the run has completed.
+        """
+        # Note: It might be nice to expose a synchronous interface for iteration, but we shouldn't do it
+        # on this class or IDEs won't warn you if you accidentally use `for` instead of `async for` to iterate.
+        return await self._graph_run.next(node)
+
+    def usage(self) -> _usage.Usage:
+        """Get usage statistics for the run so far, including token usage, model requests, and so on."""
+        return self._graph_run.state.usage
+
+    def __repr__(self):
+        final_result = self._graph_run.final_result
+        result_repr = '<run not finished>' if final_result is None else repr(final_result.result)
+        kws = [f'result={result_repr}', f'usage={self.usage()}']
+        return '<{} {}>'.format(type(self).__name__, ' '.join(kws))
```

Comment:
better to follow a dataclass and look like python code to call the class, e.g.
```suggestion
        return f'{type(self).__name__}(result={result_repr!r}, f'usage={self.usage()!r})'
```

Reply:
I purposely did not do this _because_ you can't use this python code to instantiate the class. I subscribe to the idea that whenever possible, `repr` should render as code that you can use to instantiate a (functionally equivalent) instance, but also further, that it should not _look_ like code when that isn't the case, since it will mess with your intuition / may lead you to believe that it is valid code.

Happy to change though if you insist.